write-fonts/src/pens.rs

```diff
@@ -0,0 +1,369 @@
+
+    fn close(&mut self) {
+        self.inner_pen.close();
+    }
+}
+
+pub struct BezPathPen {
+    path: BezPath,
+}
+
+fn pt(x: f32, y: f32) -> Point {
+    Point {
+        x: x as f64,
+        y: y as f64,
+    }
+}
+
+impl BezPathPen {
+    pub fn new() -> BezPathPen {
+        BezPathPen {
+            path: BezPath::new(),
+        }
+    }
+
+    pub fn into_inner(self) -> BezPath {
+        self.path
+    }
+}
+
+impl Default for BezPathPen {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Pen for BezPathPen {
+    fn move_to(&mut self, x: f32, y: f32) {
+        self.path.move_to(pt(x, y))
+    }
+
+    fn line_to(&mut self, x: f32, y: f32) {
+        self.path.line_to(pt(x, y))
+    }
+
+    fn quad_to(&mut self, cx0: f32, cy0: f32, x: f32, y: f32) {
+        self.path.quad_to(pt(cx0, cy0), pt(x, y));
+    }
+
+    fn curve_to(&mut self, cx0: f32, cy0: f32, cx1: f32, cy1: f32, x: f32, y: f32) {
+        self.path.curve_to(pt(cx0, cy0), pt(cx1, cy1), pt(x, y));
+    }
+
+    fn close(&mut self) {
+        self.path.close_path();
+    }
+}
+
+#[derive(Debug)]
+pub enum ContourReversalError {
+    InvalidFirstCommand(PenCommand),
+}
+
+/// Buffers commands until a close is seen, then plays in reverse on inner pen.
+///
+/// Reverses the winding direction of the contour. Keeps the first point unchanged.
+///
+/// <https://github.com/fonttools/fonttools/blob/78e10d8b42095b709cd4125e592d914d3ed1558e/Lib/fontTools/pens/reverseContourPen.py#L8>
+pub struct ReverseContourPen<'a, T: Pen> {
+    inner_pen: &'a mut T,
+    pending: Vec<PenCommand>,
+}
+
+fn end(subpath_start_x: f32, subpath_start_y: f32, subpath_cmd: PenCommand) -> (f32, f32) {
+    match subpath_cmd {
+        PenCommand::MoveTo { x, y } => (x, y),
+        PenCommand::LineTo { x, y } => (x, y),
+        PenCommand::QuadTo { x, y, .. } => (x, y),
+        PenCommand::CurveTo { x, y, .. } => (x, y),
+        PenCommand::Close => (subpath_start_x, subpath_start_y),
+    }
+}
+
+/// Reverse the commands in a path.
+///
+///  FontTools version in
+/// <https://github.com/fonttools/fonttools/blob/78e10d8b42095b709cd4125e592d914d3ed1558e/Lib/fontTools/pens/reverseContourPen.py#L25>
+fn flush_subpath<T: Pen>(commands: &[PenCommand], pen: &mut T) -> Result<(), ContourReversalError> {
+    if commands.is_empty() {
+        return Ok(());
+    }
+
+    let mut commands = commands;
+    let mut reversed = Vec::new();
+
+    // subpath by definition contains 0 or 1 move commands, and the move is at [0] if it exists
+    let (start_x, start_y) = if let PenCommand::MoveTo { x, y } = commands[0] {
+        // When reversed, the move is to the end point of the last command
+        // in a typical [move, ..., close] structure, end point == start point
+        let (end_x, end_y) = end(x, y, *commands.last().unwrap());
+        reversed.push(PenCommand::MoveTo { x: end_x, y: end_y });
+        commands = &commands[1..];
+        (x, y)
+    } else {
+        (0.0, 0.0)
```

Comment:
in fontTools Pen protocol, a moveTo is required all the time, it's impossible/forbidden to start a new subpath without one. Where's this (0,0) fallback coming from? We should return an an error here.

Reply:
actually, in the fonttools pen protocol, there is only one exception to this rule that all subpaths must begin with a moveTo: i.e. a contour that has no on-curve points and is only made up of off-curves from quadratic curves, it's a TrueType glyf thing. In this case the whole contour is comprised of one big qCurveTo with n>1 off-curves and the last on-curve point is set to `None` (and a closePath() to mark its end).
see https://github.com/fonttools/fonttools/blob/bf265ce49e0cae6f032420a4c80c31d8e16285b8/Lib/fontTools/pens/basePen.py#L332

However, I think in our Pen trait quad_to means a single quadratic bezier, not a fancy TrueType-style quadratic B-spline with implied on-curves, so at the moment we can't sypport that

Reply:
Assuming the pen starts at 0,0 was just an intuition really, e.g. if you just said "line to x,y" it would mean "from current, which is 0,0 by default." I'm ok with making that an error.

Reply:
Fixed.